### PR TITLE
Use device_uvector, device_span in sort groupby

### PR DESCRIPTION
**Pull request by karthikeyann**
- Replace device_vector with device_uvector
- Replace device_vector const& with device_span<const>

Ref. https://github.com/rapidsai/cudf/pull/7387#discussion_r587718013

### CODE REVIEW

File: cpp/src/groupby/sort/sort_helper.cu [hunk 1]
```diff
@@ -182,9 +182,10 @@ sort_groupby_helper::index_vector const& sort_groupby_helper::group_offsets(
       permuted_row_equality_comparator<false>(*device_input_table, sorted_order));
   }
 
-  size_type num_groups          = thrust::distance(_group_offsets->begin(), result_end);
-  (*_group_offsets)[num_groups] = num_keys(stream);
-  _group_offsets->resize(num_groups + 1);
+  size_type num_groups = thrust::distance(_group_offsets->begin(), result_end);
+  //(*)[num_groups] = num_keys(stream);
```
Comment (hyperbolic2346) [inline cpp/src/groupby/sort/sort_helper.cu:186]:
```suggestion
```

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]

**Comment by karthikeyann** [comment]
rerun tests

**Comment by jrhemstad** [comment]
CI is failing due to a segfault in a groupby test. You should verify that none of the places where `device_vector` was replaced with `device_uvector` relied on the default initialization of `device_vector`.

**Review by hyperbolic2346 — APPROVED** [review]
This looks good to me.

**Comment by jrhemstad** [comment]
rerun tests

**Review by davidwendt — APPROVED** [review]
Looks good. Some of these files need their copyright year updated.

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**